File: server/comms.js
// ===========================
//    Server heading stuff
// ===========================

let logo = '\n\t    _/_/_/_/        _/_/_/  _/    _/  _/      _/    _/_/_/        _/      _/_/_/    _/_/_/      _/_/    _/      _/    _/_/_/  _/    _/   \n'
            + '\t   _/            _/        _/    _/    _/  _/    _/            _/_/      _/    _/  _/    _/  _/    _/  _/_/    _/  _/        _/    _/    \n'
            + '\t  _/_/_/        _/  _/_/  _/    _/      _/        _/_/          _/      _/_/_/    _/_/_/    _/_/_/_/  _/  _/  _/  _/        _/_/_/_/     \n'
            + '\t       _/      _/    _/  _/    _/      _/            _/        _/      _/    _/  _/    _/  _/    _/  _/    _/_/  _/        _/    _/      \n'
            + '\t_/_/_/          _/_/_/    _/_/        _/      _/_/_/          _/      _/_/_/    _/    _/  _/    _/  _/      _/    _/_/_/  _/    _/ \n';

console.log('\033[2J'); // Clear screen
console.log('\033[0;0H'); // Move cursor to top left
console.log('\033[36m'); // Change color to blue
console.log(logo);
console.log('\033[31m'); // Change color to red
console.log('\033[37m'); // Change color to white

// ===========================
//    Get the dependencies
// ===========================

// Start express application
const app = require('express')();

const bodyParser = require('body-parser');

let server = require('http').Server(app);
let io = require('socket.io')(server);

// Support json encoded bodies
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));

// ===========================
//       Socket for app
// ===========================

// Global variable to keep a counter of the current socket connections
let currentConnections = 0;

// To execute system commands
let commandLine = require('child_process');
let runningCommand = null;

/*  Event: connection
*
*   socket.io connection event.
*   Increments a counter for the number of connections, if more than one app connects then it'll disconnect them.
*
*   Parameters:
*
*   socket - Parameter containing all the data about the socket connection.
*
*   Returns:
*
*   Nada
*
*   See Also:
*
*   <disconnect>
*/
io.on('connection', function (socket) {
    if(currentConnections >= 1)
    {
        console.log('Too many apps attempted to connect, kicked ' + socket.id);
        console.log('Current connections -> ' + currentConnections + '\n');
        io.sockets.connected[socket.id].disconnect();
    }
    else
    {
        ++currentConnections;
        console.log('App connected with id ' + socket.id);
        console.log('Current connections -> ' + currentConnections + '\n');
    }

    /*  Event: disconnect
    *
    *   socket.io disconnect event.
    *   Decrements the counter for number of connections, keeping track of the current connections.
    *
    *   Returns:
    *
    *   Nada
    *
    *   See Also:
    *
    *   <connection>
    */
    socket.on('disconnect', function () {
        --currentConnections;
        console.log('App disconnected with id ' + socket.id);
        console.log('Current connections -> ' + currentConnections + '\n');

        if(runningCommand)
        {
            console.log('Killing process with pid -> ' + (runningCommand.pid+1));
            process.kill(runningCommand.pid+1, 'SIGTERM');
        }
        });

    socket.on('message', function () {
        io.emit('Echo response');
        });

    socket.on('arm', function() {
        console.log('arming...');

        try
        {
            console.log('starting object recognition');

            process.chdir('../object-recognition/src/darknet_/');
            runningCommand = commandLine.exec('./darknet detector demo cfg/animals.data cfg/animals.cfg backup/animals_last.weights -c 0 -thresh 0.7 -json_port 8080 -out_filename data/videos/output/res.mkv > /dev/null');
            process.chdir('../../../server/');

            console.log('started object recognition with PID -> ' + runningCommand.pid);
        }
        catch (err)
        {
            console.error(err);
        }
    });

    socket.on('kill', function() {
        console.log('Killing process with pid -> ' + (runningCommand.pid+1));
        process.kill(runningCommand.pid+1, 'SIGTERM');
    });
});

// ===========================
//     Detection endpoint
// ===========================

let lastDetectedFrame = 0;
let previousDetections = [];
let newDetections = [];

/*  Endpoint: detection
*
*   Endpoint to alert the server of an object detection.
*   The object detection will send a POST request to the server notifying it of a detection.
*
*   Parameters:
*
*   req - Parameter containing all the data about the request being sent.
*   res - Parameter containing all the response data that will be sent back to the request sender.
*
*   Returns:
*
*   Success code of 200 if the function completes.
*
*   See Also:
*
*   <test>
*/
app.post('/detection', (req, res) => {
    // Test -> curl -X POST -d '{'Test':123}' -H 'Content-Type:application/json' http://127.0.0.1:8080/detection
    // ./darknet detector demo cfg/previousDetections.data cfg/previousDetections.cfg backup/previousDetections_last.weights data/videos/rhino-drone.mp4 -thresh 0.7 -json_port 8080

    /*  Data received in format
    *   JSON data:
    *   {
    *      frame_id: 39,
    *      objects:
    *      [{
    *          class_id: 2,
    *          name: 'rhino',
    *          relative_coordinates: [Object],
    *          confidence: 0.854001
    *      }]
    *   }
    */

    let detection = req.body;
    // Server will only check first detection and then each 50th frame thereafter
    if(Math.abs(detection['frame_id'] - lastDetectedFrame) > 50 || lastDetectedFrame === 0)
    {
        // If theres been over 100 frames without a detection, erase the list of old previousDetections as the camera has probably moved long past the last animal
        if(Math.abs(detection['frame_id'] - lastDetectedFrame) > 100)
        {
            previousDetections = [];
        }

        lastDetectedFrame = detection['frame_id'];

        let dX, dY;
        let threshX, threshY;
        let animalAlreadyDetected;

        // Count how many of each animal are present in the frame
        let animalCounters = [];
        for(let detectedAnimal of detection['objects'])
        {
            if(animalCounters[detectedAnimal.name])
            {
                ++animalCounters[detectedAnimal.name]['count'];
            }
            else
            {
                // Initialise the animals details, saving its name as the index, a count for how many of them there are, and its coordinates
                animalCounters[detectedAnimal.name] = [];
                animalCounters[detectedAnimal.name]['count'] = 1;
                animalCounters[detectedAnimal.name]['relative_coordinates'] = detectedAnimal['relative_coordinates'];
            }
        }

        for(let detectedAnimal in animalCounters)
        {
            // See if the old list of previousDetections animalAlreadyDetected the ones now detected
            animalAlreadyDetected = false;

            // If there is more than 1 of this type of animal, then handle it as a herd. Else check if its the same animal as previously
            if(animalCounters[detectedAnimal]['count'] > 1)
            {
                for(let animal of previousDetections)
                {
                    // If there exists a previousDetections name equal to the newly detected animal and the herd flag is set, then that animal has already been detected
                    if(animal['name'] === detectedAnimal && animal['herd'])
                    {
                        animalAlreadyDetected = true;
                        break;
                    }
                }

                if(!animalAlreadyDetected)
                {
                    console.log('New detection!');
                    console.log('\tFrame -> ' + detection['frame_id']);
                    console.log('\tHerd of -> ' + detectedAnimal);

                    io.emit('detection', { data: detection });
                }

                // Create a new list of the currently detected animals, with herd flag set to true
                newDetections.push({'name': detectedAnimal, 'relative_coordinates': animalCounters[detectedAnimal]['relative_coordinates'], 'herd': true});
            }
            else
            {
                // Check if the list of detections animalAlreadyDetected an animal with the same name, and if its position is within the threshold
                for(let animal of previousDetections)
                {
                    dX = Math.abs(animal['relative_coordinates']['center_x'] - animalCounters[detectedAnimal]['relative_coordinates']['center_x']);
                    dY = Math.abs(animal['relative_coordinates']['center_y'] - animalCounters[detectedAnimal]['relative_coordinates']['center_y']);

                    threshX = 4*animalCounters[detectedAnimal]['relative_coordinates']['width'];
                    threshY = 4*animalCounters[detectedAnimal]['relative_coordinates']['height'];

                    // If names are the same, and their X and Y coords are within a range of the thresholds, then its considered the same animal
                    if(animal['name'] === detectedAnimal && dX < threshX && dY < threshY)
                    {
                        animalAlreadyDetected = true;

                        // Update the previousDetections center coordinates
                        animal['relative_coordinates']['center_x'] = animalCounters[detectedAnimal]['relative_coordinates']['center_x'];
                        animal['relative_coordinates']['center_y'] = animalCounters[detectedAnimal]['relative_coordinates']['center_y'];
                        break;
                    }
                }

                // New detection
                if(!animalAlreadyDetected)
                {
                    console.log('New detection!');
                    console.log('\tFrame -> ' + detection['frame_id']);
                    console.log('\tAnimal -> ' + detectedAnimal);

                    io.emit('detection', { data: detection });
                }

                // Create a new list of the currently detected animals, with the herd flag set to false
                newDetections.push({'name': detectedAnimal, 'relative_coordinates': animalCounters[detectedAnimal]['relative_coordinates'], 'herd': false});
            }
        }

        // Replace the old list with the new list, since the camera is always moving 'forward', you can discard any previousDetections that have fallen out of frame
        previousDetections = newDetections;
        newDetections = [];
    }

    res.status(200).send();
});

/*  Endpoint: test
*
*   Endpoint for unit testing
*   Endpoint to test socket connection since it cannot really be sent
*
*   Parameters:
*
*   req - Parameter containing all the data about the request being sent.
*   res - Parameter containing all the response data that will be sent back to the request sender.
*
*   Returns:
*
*   Success code of 200 if the function completes
*   JSON array with a key-value pair of [status : 'connected']
*
*   See Also:
*
*   <detection>
*/
app.get('/test', (req, res) => {
    //res.write('Connection established');
    res.status(200).json({
        status : 'connected'
    });
});

// ===========================
//           Ports
// ===========================

let port = (process.env.PORT | 8080);

app.listen(port, function () {
    console.log('Server is running on port -> ' + port + '\n');
});

let socketPort = 6969;
server.listen(socketPort, function () {
    console.log('App socket listening on port -> ' + socketPort + '\n');
});

File: server/comms.py
from flask import Flask, request, jsonify
from flask_socketio import SocketIO, emit, ConnectionRefusedError
from flask_cors import CORS

from controllers.system.system_controller import SystemController
from controllers.object_detection.detection_controller import DetectionController

from controllers.object_detection.video import Video
from controllers.object_detection.webcam import Webcam

from exceptions.connection_exceptions import DroneConnectionError
from exceptions.connection_exceptions import IncorrectNetwork
from exceptions.detection_exceptions import DetectionException

from  classes.darknet_config import DarknetConfig

import subprocess
import signal, os
import glob
import base64
import time
import shlex
import threading

from controllers.system.no_drone import NoDrone
# from controllers.system.with_drone import WithDrone

# l1 = '\n\t    _/_/_/_/        _/_/_/  _/    _/  _/      _/    _/_/_/        _/      _/_/_/    _/_/_/      _/_/    _/      _/    _/_/_/  _/    _/   \n'
# l2 = '\t   _/            _/        _/    _/    _/  _/    _/            _/_/      _/    _/  _/    _/  _/    _/  _/_/    _/  _/        _/    _/    \n'
# l3 = '\t  _/_/_/        _/  _/_/  _/    _/      _/        _/_/          _/      _/_/_/    _/_/_/    _/_/_/_/  _/  _/  _/  _/        _/_/_/_/     \n'
# l4 = '\t       _/      _/    _/  _/    _/      _/            _/        _/      _/    _/  _/    _/  _/    _/  _/    _/_/  _/        _/    _/      \n'
# l5 = '\t_/_/_/          _/_/_/    _/_/        _/      _/_/_/          _/      _/_/_/    _/    _/  _/    _/  _/      _/    _/_/_/  _/    _/ \n'
# logo = l1 + l2 + l3 + l4 + l5

l1 = "\t   dvvvvvo.   v vvvvvvvvvv v vvvvvvvo. `v.`vvvb         ,v' v vvvvvvvvvv v vvvvvvvo.    b.             v         .v.   vvvvvvv vvvvvvvv  ,ovvvvvvo.    v vvvvvvvo.\n"
l2 = "\t v.`vvv.  Yv  v vvv        v vv     `vv `v.`vvvb       ,v'  v vvv        v vv     `vv   Yvvvvo.        v        :vvv.        v vv     ,v vv      `vb   v vv     `vv\n"
l3 = "\t `v.`vvv.     v vvv        v vv     ,vv  `v.`vvvb     ,v'   v vvv        v vv     ,vv   .`Yvvvvvo.     v       .`vvvv.       v vv     vv vv       `vb  v vv     ,vv\n"
l4 = "\t  `v.`vvv.    v vvvvvvvvvv v vv.   ,vv'   `v.`vvvb   ,v'    v vvvvvvvvvv v vv.   ,vv'   vo. `Yvvvvvo.  v      .v.`vvvv.      v vv     vv vv        vv  v vv.   ,vv'\n"
l5 = "\t   `v.`vvv.   v vvv        v vvvvvvvP'     `v.`vvvb ,v'     v vvv        v vvvvvvvP'    v`Yvo. `Yvvvvoov     .v`v.`vvvv.     v vv     vv vv        vv  v vvvvvvvP'\n"
l6 = "\t    `v.`vvv.  v vvv        v vv`vb          `v.`vvvbv'      v vvv        v vv`vb        v   `Yvo. `Yvvvv    .v' `v.`vvvv.    v vv     vv vv        vP  v vv`vb\n"
l7 = "\tvb   `v.`vvv. v vvv        v vv `vb.         `v.`vvv'       v vvv        v vv `vb.      v      `Yvo. `Yv   .v'   `v.`vvvv.   v vv     `v vv       ,vP  v vv `vb.\n"
l8 = "\t`vb.  ;v.`vvv v vvv        v vv   `vb.        `v.`v'        v vvv        v vv   `vb.    v         `Yvo.`  .vvvvvvvvv.`vvvv.  v vv     `v vv     ,vv'   v vv   `vb.\n"
l9 = "\t `YvvvP ,vvP' v vvvvvvvvvv v vv     `vv.       `v.`         v vvvvvvvvvv v vv     `vv.  v            `Yo .v'       `v.`vvvv. v vv       `vvvvvvvP'     v vv     `vv.\n"
logo = l1 + l2 + l3 + l4 + l5 + l6 + l7 + l8 + l9

print('\033[2J') # Clear screen
print('\033[00H') # Move cursor to top left
print('\033[36m') # Change color to blue
print(logo)
# print('\033[31m') # Change color to red
print('\033[37m') # Change color to white

# ============================================================================
#                           Setup of server
# ============================================================================



# Port for the server
port = 42069
host = '0.0.0.0'

# Flask app
app = Flask(__name__)
CORS(app)

# Create the socket, with all origins allowed
io = SocketIO(app, cors_allowed_origins="*", monitor_clients=True)
# Darknet configs
yolov3 = DarknetConfig(weights='yolov3.weights', cfg='yolov3.cfg', data='coco.data')
animals = DarknetConfig(weights='animals-tiny_last.weights', cfg='animals-tiny.cfg', data='animals.data')
yolov3_http = DarknetConfig(weights='yolov3.weights', cfg='yolov3.cfg', data='coco.data', url='http://192.168.8.101:4747/mjpegfeed')

# systemContoller = SystemController( WithDrone() ) # Controller with the drone
# systemContoller = SystemController( NoDrone( Webcam(camera_id=2) ) ) # Controller using webcam 0
systemContoller = SystemController( NoDrone( Video(config=yolov3_http) ) ) # Controller using video + pass in config param

# Create a detection controller that can use the io object
detectionController = DetectionController(io)

# ============================================================================
#                           Socket for the app
# ============================================================================

@io.on('connect')
def connect():
    try:
        systemContoller.connectDrone()
        print('App connected with ID: ', request.sid, '\n')
    except DroneConnectionError as error:
        print( error.getMessage() )
        raise ConnectionRefusedError('Failed')
    except IncorrectNetwork as error:
        print( error.getMessage() )

# Disconnection event
@io.on('disconnect')
def disconnect():
    systemContoller.disconnectDrone()
    print('App disconnected with ID', request.sid)

# Arming event
@io.on('arm_drone')
def arm():
    systemContoller.armDrone()
    emit('drone_armed')

# Disarm event
@io.on('disarm_drone')
def disarm():
    systemContoller.disarmDrone()
    emit('drone_disarmed')

# Default GET, should never happen
@app.route('/', methods=['GET'])
def index():
    return '<html><head><title>Turn back now</title></head><body><p style="color: red; width: 100%; text-align: center; margin-top: 20%">01011001011011110111010100100000011100110110100001101111011101010110110001100100011011100010011101110100001000000110001001100101001000000110100001100101011100100110010100100001</p></body></html>', 200

# Endpoint to ping server
@app.route('/ping', methods=['GET'])
def ping():
    return '[{"pong"}]', 200

# Update coords event
@app.route('/coords', methods=['POST'])
def coords():
    # global bebop

    # print('Changing location')

    # percentage = bebop.getBatteryPercentage()
    # print('\nBattery percentage: {}%'.format(percentage))

    # # make sure the drone is initialized
    # if percentage != 'uninitialized':
    #     percentage = int(percentage)

    #     # If the battery percentage is below 15% then return home and land
    #     if percentage <= 15:
    #         bebop.go_home()
    #         bebop.land_drone()

    #         print('Battery low, returning home')
    #         io.emit('battery-low')
    #     else:
    #         print('Received:', request.get_json())

    #         lat = request.get_json()['lat']
    #         lon = request.get_json()['lon']

    #         print('Moving to:', lat, lon)
    #         bebop.newCenterLocation(lat, lon)
    # else:
    #     print('uninitialized')

    return '', 200

# ============================================================================
#                             Handling images
# ============================================================================

# Converts the image in detections/ to base64
def convertImageToBase64(image_id):
    session_time = systemContoller.getSessionTime()
    img = '../object-recognition/detections/' +  session_time  + '/' + image_id

    with open(img, 'rb') as img_file:
        return str(base64.b64encode(img_file.read()))

# Endpoint for the app to request an image based on an image_id
@app.route('/image', methods=['POST'])
def return_image():
    blob = convertImageToBase64(request.get_json()['image'])
    
    return (jsonify(blob), 200)

# ============================================================================
#                           Handling detections
# ============================================================================

# Endpoint for POST requests alerting the server of a detection
    # Tests:
    # Single animal -> curl -X POST -d '{"frame_id":121, "objects": [ {"class_id":1, "name":"elephant", "relative_coordinates":{"center_x":0.465886, "center_y":0.690794, "width":0.048322, "height":0.065592}, "confidence":0.704248}]}' -H 'Content-Type:application/json' http://127.0.0.1:42069/detection
    # Herd -> curl -X POST -d '{"frame_id":181, "objects": [ {"class_id":1, "name":"elephant", "relative_coordinates":{"center_x":0.465886, "center_y":0.690794, "width":0.048322, "height":0.065592}, "confidence":0.704248}, {"class_id":1, "name":"elephant", "relative_coordinates":{"center_x":0.465886, "center_y":0.690794, "width":0.048322, "height":0.065592}, "confidence":0.704248}]}' -H 'Content-Type:application/json' http://127.0.0.1:42069/detection

    #  Data received in format
    #   JSON data:
    #   {
    #      frame_id: 39,
    #      objects:
    #      [{
    #          class_id: 2,
    #          name: 'rhino',
    #          relative_coordinates: [Object],
    #          confidence: 0.854001
    #      }]
    #   }
    #

@app.route('/detection', methods=['POST'])
def detection():
    detectionController.detectionFilter(request.json)

    return '', 200

# ============================================================================
#                           run socket/server
# ============================================================================

def zipdir(session_dir, password):
    subprocess.call(['7z', 'a', '-mx=9', '-t7z', '-p' + password, '-y', session_dir + '-detections.zip', session_dir + '/*'], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)


def shutdown_process():
    global bebop

    try:
        session_time = systemContoller.getSessionTime()
    except:
        session_time = False

    print('Beginning shutdown process... \n')

    systemContoller.disarmDrone()
    systemContoller.disconnectDrone()

    if session_time:
        # Create an encrypted zip file of all the detections
        print('Encrypting all detection images... ', end='', flush=True)

        os.chdir('../object-recognition/detections')
        
        # Create an encrypted zip of the files
        zipdir(session_time, 'Capstone69')

        print('Done!')

        # Delete all the unencrypted images
        print('Deleting all unencrypted detection images... ', end='', flush=True)

        os.chdir(session_time)

        files=glob.glob('*.jpg')
        for filename in files:
            os.remove(filename)

        os.chdir('../')

        os.rmdir(session_time)

        print('Done!')

        os.chdir('../../server/')

    # Stop the server
    io.emit('disconnect')
    io.stop()
    print('Done... Goodbye!')

def run(p = port, h = host):
    # Run the flask API
    print('Server running on https://' + h + ':' + str(p), '\n')

    try:
        io.run(app, port = p, host = h)
    except KeyboardInterrupt:
        print('^C received, shutting down the server...')
        
    shutdown_process()

# ============================================================================
#       Start on default port or on the one passed in as an argument
# ============================================================================
if __name__ == "__main__":
    from sys import argv

    if len(argv) == 2:
        run(p = int(argv[1]))
    else:
        run()
